File: compatibility-tests/sdk-test-java/src/test/java/io/floci/gcp/test/GcsComposeTest.java
package io.floci.gcp.test;

import com.google.cloud.storage.Blob;
import com.google.cloud.storage.BlobId;
import com.google.cloud.storage.BlobInfo;
import com.google.cloud.storage.BucketInfo;
import com.google.cloud.storage.Storage;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Regression coverage for issue #1: object compose was broken
 * ("400 Unsupported method override: null") because the route was mapped to
 * {@code :compose} instead of {@code /compose}.
 */
class GcsComposeTest {

    private static final String BUCKET_NAME = TestFixtures.uniqueName("compose-bucket");

    private static Storage storage;

    @BeforeAll
    static void setUp() {
        storage = TestFixtures.storageClient();
        storage.create(BucketInfo.of(BUCKET_NAME));
    }

    @AfterAll
    static void tearDown() throws Exception {
        if (storage instanceof AutoCloseable closeable) {
            closeable.close();
        }
    }

    @Test
    void composeConcatenatesSourceObjects() {
        storage.create(BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, "p1")).build(),
                "hello ".getBytes(StandardCharsets.UTF_8));
        storage.create(BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, "p2")).build(),
                "world".getBytes(StandardCharsets.UTF_8));

        BlobInfo target = BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, "composed"))
                .setContentType("text/plain")
                .build();
        Storage.ComposeRequest request = Storage.ComposeRequest.newBuilder()
                .addSource("p1")
                .addSource("p2")
                .setTarget(target)
                .build();

        Blob composed = storage.compose(request);

        assertThat(composed.getName()).isEqualTo("composed");
        assertThat(composed.getContentType()).isEqualTo("text/plain");

        String content = new String(storage.readAllBytes(BlobId.of(BUCKET_NAME, "composed")),
                StandardCharsets.UTF_8);
        assertThat(content).isEqualTo("hello world");
        assertThat(composed.getSize()).isEqualTo("hello world".getBytes(StandardCharsets.UTF_8).length);
    }
}

File: compatibility-tests/sdk-test-node/tests/gcs-compose.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { Storage } from '@google-cloud/storage';
import { ENDPOINT, PROJECT_ID, uniqueName } from './setup';

// Regression coverage for issue #1: object compose was broken
// ("400 Unsupported method override: null"). bucket.combine() must work.
describe('GCS compose', () => {
  let storage: Storage;
  let bucketName: string;

  beforeAll(async () => {
    storage = new Storage({ apiEndpoint: ENDPOINT, projectId: PROJECT_ID });
    bucketName = uniqueName('compose-bucket');
    await storage.createBucket(bucketName);
  });

  afterAll(async () => {
    const bucket = storage.bucket(bucketName);
    for (const name of ['p1', 'p2', 'composed']) {
      await bucket.file(name).delete().catch(() => {});
    }
    await bucket.delete().catch(() => {});
  });

  it('concatenates source objects into the destination', async () => {
    const bucket = storage.bucket(bucketName);
    await bucket.file('p1').save('hello ', { contentType: 'text/plain' });
    await bucket.file('p2').save('world', { contentType: 'text/plain' });

    await bucket.combine(['p1', 'p2'], 'composed');

    const [content] = await bucket.file('composed').download();
    expect(content.toString()).toBe('hello world');

    const [metadata] = await bucket.file('composed').getMetadata();
    expect(Number(metadata.size)).toBe(Buffer.byteLength('hello world'));
  });
});
